### PR TITLE
SoA range checking: adds inter operability between range checked and non range checked

### PR DESCRIPTION
**Pull request by ericcano**
#### PR description:

This PR adds copy and construction compatibility between `View`s and `ConstView`s (respectively, no const evasion was created) with different template parameters. This allows compile time switching to the range checking variant for debugging. The `View`s are usually used from the `PortableCollections` (and are non range checking). Changing the signature or a function or kernel to use the range checking variant simply requires a type definition change and the creation of a translating variable at the kernel call. The program will then half immediately on range violation.

#### PR validation:

The copy and construction are now validated in a test, as part of this PR. I addition this feature was used and validated to debug an actual out of range error during module debugging.

### CODE REVIEW

File: DataFormats/SoATemplate/test/SoALayoutAndView_t.hip.cc [hunk 1]
```diff
@@ -123,6 +125,12 @@ int main(void) {
   hipCheck(hipMallocHost((void**)&h_buf, hostDeviceSize));
   SoAHostDeviceLayout h_soahdLayout(h_buf, numElements);
   SoAHostDeviceView h_soahd(h_soahdLayout);
+
+  // Validation of range checking variants initialization
+  SoAHostDeviceRangeCheckingView h_soahdrc(h_soahdLayout);
+  [[maybe_unused]] SoAHostDeviceRangeCheckingView h_soahdrc2 = h_soahdLayout;
+  [[maybe_unused]] SoAHostDeviceRangeCheckingView h_soahdrc3{h_soahd};
+  [[maybe_unused]] SoAHostDeviceRangeCheckingView h_soahdrc4 = h_soahd;
```
Comment (makortel) [inline DataFormats/SoATemplate/test/SoALayoutAndView_t.hip.cc:133]: I'd suggest to add some tests on the left-hand-side objects as well to ensure the behavior of the "copy" constructor is correct.

**Comment by fwyzard** [comment]
@ericcano can we take this opportunity to change
https://github.com/cms-sw/cmssw/blob/80e1b48540f04fd1c72a74c5d058406b654474da/DataFormats/SoATemplate/interface/SoACommon.h#L54-L62

to remove `CMS_SOA_BYTE_SIZE_TYPE` and use `cms::soa::byte_size_type` everywhere ?

File: DataFormats/SoATemplate/test/SoALayoutAndView_t.hip.cc
[the remark this answers is not in the document]
Comment (ericcano) [reply DataFormats/SoATemplate/test/SoALayoutAndView_t.hip.cc:1]: We can emulate something similar, yes.

File: DataFormats/SoATemplate/test/SoALayoutAndView_t.hip.cc (hunk 1 above)
Comment (ericcano) [reply DataFormats/SoATemplate/test/SoALayoutAndView_t.hip.cc:133]: Sure!

**Comment by fwyzard** [comment]
enable gpu

**Comment by fwyzard** [comment]
please test

**Comment by ericcano** [comment]
I did not manage to reproduce the failures seen in Jenkins (I could run successfully some cmsRun tests that had failed for Jenkins, after recompiling anything the depended on SoA and Portable collections), but did notice I had switched from `size_t` to `std::size_t`, and that might be confusing ROOT(?). 

I reverted this, could we run the Jenkins tests again, @fwyzard?

**Comment by makortel** [comment]
@cmsbuild, please test

File: DataFormats/SoATemplate/interface/SoALayout.h [hunk 2]
```diff
@@ -424,7 +424,7 @@
  */
 // clang-format off
 #define GENERATE_SOA_LAYOUT(CLASS, ...)                                                                                \
-  template <CMS_SOA_BYTE_SIZE_TYPE ALIGNMENT = cms::soa::CacheLineSize::defaultSize,                                   \
+  template <cms::soa::size_type ALIGNMENT = cms::soa::CacheLineSize::defaultSize,                                      \
```
Comment (makortel) [inline DataFormats/SoATemplate/interface/SoALayout.h:427]: Not sure if this has any relation to the dictionary problem, but should this be
```suggestion
  template <cms::soa::byte_size_type ALIGNMENT = cms::soa::CacheLineSize::defaultSize,                                      \
```
? (currently this PR changes the underlying type from `std::size_t` to `cms_int32_t`)

**Comment by smuzaffar** [comment]
Milestone for this pull request has been moved to CMSSW_14_0_X.Please open a backport if it should also go in to CMSSW_13_3_X.

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by fwyzard** [comment]
enable gpu

**Comment by fwyzard** [comment]
please test

**Comment by rappoccio** [comment]
+1

**Pull request merged**